### PR TITLE
Refactor New Project Wizard to extensions to support other platforms

### PR DESCRIPTION
**Pull request by zhangzhx**
such as WebStorm.

## Description

* Added a new extension point `sam.project` for different `RuntimeGroup`
* Moved `SamProjectTemplate` management to the extension point
* Moved out the property `Sdk` to the extension point to be provided by different `RuntimeGroup`
* Moved `SdkSelectionPanel` to be managed by the extension point
* Moved `SdkSelectionPanel` to `SamInitSelectionPanel` to better encapsulate it.
* Some other small cleanups

## Motivation and Context

In order to support node.js new project wizard, we need to move all the language related components in an extension point, so that we can provide the implementation of node.js in other module.

## Related Issue(s)
#1033 

## Checklist

- [ ] I have read the **README** document
- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * Return an instance of UI section for selecting SDK for the [RuntimeGroup]
+     */
+    fun createSdkSelectionPanel(generator: SamProjectGenerator): SdkSelectionPanel
+
+    companion object : RuntimeGroupExtensionPointObject<SamProject>(ExtensionPointName("aws.toolkit.lambda.sam.project"))
+}
+
+class SamNewProjectSettings {
+    lateinit var runtime: Runtime
```
[the remark this answers is not in the document]
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt]: SamProjectGenerator can own the panel

File: jetbrains-core/resources/META-INF/ext-java.xml [hunk 2]
```diff
@@ -7,5 +7,6 @@
         <builder runtimeGroup="JAVA" implementation="software.aws.toolkits.jetbrains.services.lambda.java.JavaLambdaBuilder"/>
         <sam.debugSupport runtimeGroup="JAVA" implementation="software.aws.toolkits.jetbrains.services.lambda.java.JavaSamDebugSupport"/>
         <handlerResolver runtimeGroup="JAVA" implementation="software.aws.toolkits.jetbrains.services.lambda.java.JavaLambdaHandlerResolver"/>
+        <sam.project runtimeGroup="JAVA" implementation="software.aws.toolkits.jetbrains.services.lambda.java.JavaSamProject"/>
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/resources/META-INF/ext-java.xml]: Added `Wizard` suffix for all `SamProject`

* JavaSamProject -> JavaSamProjectWizard
* PythonSamProject -> PythonSamProjectWizard
* SamProject -> SamProjectWizard

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
+import software.aws.toolkits.jetbrains.ui.wizard.AwsModuleType
+import software.aws.toolkits.jetbrains.ui.wizard.SamInitRunner
+import software.aws.toolkits.jetbrains.ui.wizard.SamProjectGenerator
+import software.aws.toolkits.jetbrains.ui.wizard.SdkSelectionPanel
+
+/**
+ * Used to manage SAM project information for different [RuntimeGroup]s
+ */
+interface SamProject {
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt]: Renamed to `SamProjectWizard`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    lateinit var sdkSettings: SdkSettings
+}
+
+interface SdkSettings
+
+/**
+ * Sdk settings that supports [Sdk] as the language's SDK, such as Java, Python.
+ */
+class SdkBasedSdkSettings(
+    var sdk: Sdk? = null
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt]: Changed to `data class`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            location(),
+            dependencyManager()
+        )
+    }
+
+    protected open fun location(): String? = null
+
+    protected open fun dependencyManager(): String? = null
+
+    open fun supportedRuntimes(): Set<Runtime> = Runtime.knownValues().toSet()
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt]: Done.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/wizard/IntelliJSdkSelectionPanel.kt [hunk 6]
```diff
@@ -62,8 +37,26 @@ class IntelliJSdkSelectionPanel(generator: SamProjectGenerator) : SdkSelectionPa
         return null
     }
 
-    override fun getSdk(): Sdk? {
+    override fun ensureSdk() {
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/ui/wizard/IntelliJSdkSelectionPanel.kt]: The `ensureSdk` was to set the SDK to the model, ie `SamNewProjectSettings`. As we made it immutable, removed this method and created `getSdkSettings` instead to always return the update-to-date UI settings.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamProject.kt]: Removed all the `lateinit` and moved the `SamNewProjectSettings` out of the generator.

I still kept the Panel in the Peer as it was meant to manage the UIs.

**Comment by rli** [comment]
Looks fine to me; don't see anything egregious. Please see if @abrooksv has further comments.

**Pull request merged**